lib/mayaUsd/fileio/utils/meshWriteUtils.cpp

```diff
@@ -0,0 +1,380 @@
+
+            // Translate OSD2 values to OSD3.
+            if (sdFVInterpBound == UsdGeomTokens->bilinear) {
+                sdFVInterpBound = UsdGeomTokens->all;
+            } else if (sdFVInterpBound == UsdGeomTokens->edgeAndCorner) {
+                sdFVInterpBound = UsdGeomTokens->cornersPlus1;
+            } else if (sdFVInterpBound == _meshTokens->alwaysSharp) {
+                sdFVInterpBound = UsdGeomTokens->boundaries;
+            } else if (sdFVInterpBound == UsdGeomTokens->edgeOnly) {
+                sdFVInterpBound = UsdGeomTokens->none;
+            }
+        } else {
+            plug = mesh.findPlug(MString("rman__torattr___subdivFacevaryingInterp"));
+            if (!plug.isNull()) {
+                switch(plug.asInt()) {
+                    case 0:
+                        sdFVInterpBound = UsdGeomTokens->all;
+                        break;
+                    case 1:
+                        sdFVInterpBound = UsdGeomTokens->cornersPlus1;
+                        break;
+                    case 2:
+                        sdFVInterpBound = UsdGeomTokens->none;
+                        break;
+                    case 3:
+                        sdFVInterpBound = UsdGeomTokens->boundaries;
+                        break;
+                    default:
+                        break;
+                }
+            }
+        }
+
+        return sdFVInterpBound;
+    }
+
+} // anonymous namespace
+
+bool
+UsdMayaMeshUtil::getMeshNormals(const MFnMesh& mesh,
+                                VtArray<GfVec3f>* normalsArray,
+                                TfToken* interpolation)
+{
+    MStatus status{MS::kSuccess};
+
+    // Sanity check first to make sure we can get this mesh's normals.
+    const int numNormals = mesh.numNormals(&status);
+    if (status != MS::kSuccess || numNormals == 0) {
+        return false;
+    }
+
+    // Using itFV.getNormal() does not always give us the right answer, so
+    // instead we have to use itFV.normalId() and use that to index into the
+    // normals.
+    MFloatVectorArray mayaNormals;
+    status = mesh.getNormals(mayaNormals);
+    if (status != MS::kSuccess) {
+        return false;
+    }
+
+    const unsigned int numFaceVertices = mesh.numFaceVertices(&status);
+    if (status != MS::kSuccess) {
+        return false;
+    }
+
+    normalsArray->resize(numFaceVertices);
+    *interpolation = UsdGeomTokens->faceVarying;
+
+    MItMeshFaceVertex itFV(mesh.object());
+    unsigned int fvi = 0;
+    for (itFV.reset(); !itFV.isDone(); itFV.next(), ++fvi) {
+        int normalId = itFV.normalId();
+        if (normalId < 0 ||
+                static_cast<size_t>(normalId) >= mayaNormals.length()) {
+            return false;
+        }
+
+        MFloatVector normal = mayaNormals[normalId];
+        (*normalsArray)[fvi][0] = normal[0];
+        (*normalsArray)[fvi][1] = normal[1];
+        (*normalsArray)[fvi][2] = normal[2];
+    }
```

Comment:
You might want to use 
```
        MIntArray normalCounts, normalIndices;
        mesh.getNormalIds(normalCounts, normalIndices);
```
Rather than using the MItMeshFaceVertex  iterator :(

I'll direct you towards this:

https://github.com/Autodesk/maya-usd/blob/1737b11e03a37c599681b4a8802b417f0e34fb1f/plugin/al/usdmayautils/AL/usdmaya/utils/MeshUtils.cpp#L2180

That's the approach we've taken in the AL plugin (which also generates the perVert/perFace interp modes, and also has the option to write the normals out as a primvar,  which we need in a couple of places in our pipeline)

Reply:
@robthebloke see 2b1e304
Yes, I have already seen some old codes in Pixar code base that definitely has room for improvement.